packages/navigator/src/browser/open-editors-widget/navigator-open-editors-tree-model.ts

```diff
@@ -0,0 +1,112 @@
+import { ApplicationShell, CompositeTreeNode, Navigatable, SelectableTreeNode, Widget } from '@theia/core/lib/browser';
+import { EditorWidget } from '@theia/editor/lib/browser';
+import { WorkspaceService } from '@theia/workspace/lib/browser';
+import { EditorPreviewManager, EditorPreviewWidget } from '@theia/editor-preview/lib/browser';
+import { DisposableCollection } from '@theia/core/lib/common';
+import debounce = require('@theia/core/shared/lodash.debounce');
+
+export interface OpenEditorNode extends FileStatNode {
+    widget: Widget;
+};
+
+export namespace OpenEditorNode {
+    export function is(node: object | undefined): node is OpenEditorNode {
+        return FileStatNode.is(node) && 'widget' in node;
+    }
+}
+
+@injectable()
+export class OpenEditorsModel extends FileTreeModel {
+    @inject(ApplicationShell) protected readonly applicationShell: ApplicationShell;
+    @inject(WorkspaceService) protected readonly workspaceService: WorkspaceService;
+    @inject(EditorPreviewManager) protected readonly editorPreviewManager: EditorPreviewManager;
+
+    protected toDisposeOnPreviewWidgetReplaced = new DisposableCollection();
+
+    @postConstruct()
+    protected init(): void {
+        super.init();
+        this.initializeRoot();
+    }
+
+    protected async initializeRoot(): Promise<void> {
+        this.selectionService.onSelectionChanged(selection => {
+            const { widget } = (selection[0] as OpenEditorNode);
+            this.applicationShell.activateWidget(widget.id);
+        });
+        this.toDispose.push(this.applicationShell.onDidChangeCurrentWidget(async ({ newValue }) => {
+            const nodeToSelect = this.tree.getNode(newValue?.id) as SelectableTreeNode;
+            if (nodeToSelect) {
+                this.selectNode(nodeToSelect);
+            }
+        }));
+        this.toDispose.push(this.workspaceService.onWorkspaceChanged(() => this.updateOpenWidgets()));
+        this.toDispose.push(this.workspaceService.onWorkspaceLocationChanged(() => this.updateOpenWidgets()));
+        this.toDispose.push(this.applicationShell.onDidAddWidget(() => this.updateOpenWidgets()));
+        this.toDispose.push(this.applicationShell.onDidRemoveWidget(() => this.updateOpenWidgets()));
+        this.toDispose.push(this.editorPreviewManager.onCreated(previewWidget => {
+            if (previewWidget instanceof EditorPreviewWidget) {
+                this.toDisposeOnPreviewWidgetReplaced.dispose();
+                this.toDisposeOnPreviewWidgetReplaced.push(previewWidget.onPinned(() => this.updateOpenWidgets()));
+                this.toDisposeOnPreviewWidgetReplaced.push(previewWidget.onDidChangeTrackableWidgets(() => this.updateOpenWidgets()));
+            }
+        }));
+        await this.updateOpenWidgets();
+        this.fireChanged();
+    }
+
+    protected updateOpenWidgets = debounce(this.doUpdateOpenWidgets, 250);
+
+    protected async doUpdateOpenWidgets(): Promise<void> {
```

Comment:
Would there be any value in a more surgical approach here, adding / removing nodes as widgets are opened / closed, rather than rebuilding the tree completely for every widget opening or closing?

Reply:
I thought about this as well, but there were a lot of strange timing issues that I came across and it seemed quite difficult to validate removing/adding individual nodes reliably, especially when accounting for the PreviewWidget. In the end I found it easier to rebuild. If you have any suggestions I'm happy to discuss